walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * We loop through all given events to find all available TimeRanges. Pointer 
+      * rangeStart is pointing at the start of the current available TimeRange, which 
+      * is the time between the pointer and start time of the current event.If the current 
+      * available TimeRange has longer duration than the requested meeting duration, add
+      * it to available meetingTimes. Then rangeStart is updated to the end of the current
+      * event.
+      */
+    private static Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events, Collection<String> attendees, long duration) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        // Sort the events by the start time.
+        List<Event> eventsSorted = new ArrayList<Event>(events);
+        Collections.sort(eventsSorted, Event.SORT_BY_START_ASCENDING);
+
+        int rangeStart = DAY_START; // Pointer to start of the previous event.
+
+        // Linear search the current events and add new TimeRange to 
+        // meetingTimes.
+        for (Event event : eventsSorted) {
+            Collection<String> eventAttendees = event.getAttendees();
+            TimeRange eventWhen = event.getWhen();
+            int eventStart = eventWhen.start();
+            int eventEnd = eventWhen.end();
+
+            // Check if given attendees shows up in the event.
+            if (isAttending(attendees, eventAttendees)) {
+                // Check if the current event starts before the previous event ends,
+                // or there is no gap in between.
+                if (rangeStart >= eventStart) {
+                    // If so, check which ends earilier and change the start of next 
+                    // available TimeRange to the later end time of the two.
+                    if (eventEnd > rangeStart) {
+                        rangeStart = eventEnd;
+                    }
+                } else {
+                    // If they don't overlap, initialize the current available TimeRange.            
+                    TimeRange newRange = TimeRange.fromStartEnd(
+                            rangeStart, eventStart, false);
+                        if (newRange.duration() >= duration) {
+                            meetingTimes.add(newRange);
+                        }
+                    rangeStart = eventEnd;
+                }
+            }
+        }
+
+        TimeRange finalRange = TimeRange.fromStartEnd(rangeStart, DAY_END, true);
+        if (finalRange.duration() >= duration) {
+            meetingTimes.add(finalRange);
+        }
+
+        return meetingTimes;
+    }
+
+    /**
+      * Check if any eventAttendees present in the given collection of attendees. 
+      * Return true if any given attendees present in the current event. Return
```

Comment:
I won't say "current event" since isAttending() doesn't know about event. It only knows 'attendees and eventAttendees'. I'd replace the "current event" with "eventAttendees".

Reply:
The comment is updated! :)